flatsurf/geometry/homology.py

```diff
@@ -0,0 +1,895 @@
+        r"""
+        Normalize parameters used to construct homology.
+
+        TESTS:
+
+        Homology is unique and cached::
+
+            sage: from flatsurf import translation_surfaces, SimplicialHomology
+            sage: T = translation_surfaces.torus((1, 0), (0, 1))
+            sage: T.set_immutable()
+            sage: SimplicialHomology(T) is SimplicialHomology(T)
+            True
+
+        """
+        # TODO: Change default implementation to spanning_tree
+        if surface.is_mutable():
+            raise ValueError("surface must be immutable to compute homology")
+
+        from sage.all import ZZ
+        coefficients = coefficients or ZZ
+
+        # TODO: Use a better category.
+        from sage.all import Sets
+        category = category or Sets()
+        subset = frozenset(subset or {})
+
+        return super().__classcall__(cls, surface, coefficients, generators, subset, implementation, category)
+
+    def __init__(self, surface, coefficients, generators, subset, implementation, category):
+        Parent.__init__(self, category=category)
+
+        from sage.categories.all import Rings
+        if coefficients not in Rings():
+            raise TypeError("coefficients must be a ring")
+
+        if generators not in ["edge",]:
+            raise NotImplementedError("cannot represented homology with these generators yet")
+
+        if subset:
+            raise NotImplementedError("cannot compute relative homology yet")
+
+        if implementation not in ["generic", "spanning_tree"]:
+            raise NotImplementedError("cannot compute homology with this implementation yet")
+
+        self._surface = surface
+        self._coefficients = coefficients
+        self._generators = generators
+        self._subset = subset
+        self._implementation = implementation
+
+    def surface(self):
+        r"""
+        Return the surface of which this is the homology.
+
+        EXAMPLES::
+
+            sage: from flatsurf import translation_surfaces, SimplicialHomology
+            sage: T = translation_surfaces.torus((1, 0), (0, 1))
+            sage: T.set_immutable()
+            sage: H = SimplicialHomology(T)
+            sage: H.surface() == T
+            True
+
+        """
+        return self._surface
+
+    def matrix(self, deformation, homology=None):
+        r"""
+        Return a square matrix representing the ``deformation`` on the
+        :meth:`surface` by writing the images of the generators of this
+        homology in terms of the images of the ``homology`` of the codomain of
+        ``deformation``.
+
+        INPUT:
+
+        - ``homology`` -- a :class:`SimplicialHomology` (default: ``None``); if
+          ``None``, the homology of the codomain is determined automatically.
+
+        EXAMPLES::
+
+            sage: from flatsurf import translation_surfaces, SimplicialHomology
```

Comment:
@sfreedman67 This works now (somewhat). I have not checked the outputs at all, so this might be completely wrong somehow.

There's a bug which will be fixed in https://github.com/flatsurf/flatsurf/pull/328. If you get an error that no isomorphism could be found, then that's to blame.

To see the generators of homology use `H.gens()`. For the L this gives:

```
(B[(0, 1)], B[(0, 0)], B[(1, 1)], B[(2, 0)])
```

So the generators are the edge 1 of polygon 0, edge 0 of polygon 0, edge 1 of polygon 1, and edge 0 of polygon 2.

Reply:
I'm excited to test it out!